File: src/main_menu.rs
use game::GameState;
use geometry::*;
use gui::*;
use io::*;
use states::*;

#[allow(dead_code)]
pub struct MainMenuState
{
    // *** ONLY FOR DEMO PURPOSES - IMPLEMENT A BETTER SETUP ***
    new_game_button: Button,
}

impl MainMenuState
{
    pub fn new() -> MainMenuState
    {
        MainMenuState {
            new_game_button: Button::new_wide(P { x: 300, y: 128 }, "New game"),
        }
    }
}

impl State for MainMenuState
{
    fn name(&self) -> &str
    {
        return "Main menu";
    }

    fn on_pushed(&mut self)
    {
    }

    fn on_start(&mut self) -> Vec<StateSignal>
    {
        return Vec::new();
    }

    fn on_resume(&mut self)
    {
    }

    fn draw(&mut self, io: &mut Io)
    {
        // *** ONLY FOR DEMO PURPOSES - IMPLEMENT A BETTER SETUP ***

        io.draw_text(
            "This is the main menu.",
            64,
            64,
            TextSize::Big,
            TextAnchorX::Left,
            TextAnchorY::Top,
        );

        self.new_game_button.draw(io);

        io.draw_tile(P::new(0, 0), P::new(32, 400));
        io.draw_tile(P::new(32, 0), P::new(64, 400));
        io.draw_tile(P::new(64, 0), P::new(96, 400));
    }

    fn update(&mut self, io: &mut Io) -> Vec<StateSignal>
    {
        let d = io.read();

        // *** ONLY FOR DEMO PURPOSES - IMPLEMENT A BETTER SETUP ***
        self.new_game_button.update(&d);

        // Proceed to game state?
        if self.new_game_button.is_triggered || d.char == 'n' {
            let game_state = Box::new(GameState::new());

            return vec![StateSignal::Push { state: game_state }];
        }

        // Exit game?
        if d.key == Key::Esc {
            return vec![StateSignal::Pop];
        }

        return Vec::new();
    }

    fn on_popped(&mut self)
    {

    }
}

File: src/main_menu.cpp
#include "main_menu.hpp"

#include "io.hpp"
#include "game.hpp"

std::vector<StateSignal> MainMenu::on_start()
{

        {
                Button button("Start", P(32+11, 32+11), P(64, 32), {10, 213, 10}, {213, 100, 10});
                
                buttons_.push_back(std::move(button));
        }

        {
                Button button("Quit", P(32+11, 66+11), P(64, 32), {10, 213, 10}, {213, 100, 10});
                
                buttons_.push_back(std::move(button));
        }
        
        return {};
}

void MainMenu::draw()
{
        for (auto& elem : buttons_)
        {
                elem.draw();
        }    
}

std::vector<StateSignal> MainMenu::update(const InputData& input)
{
        for (auto& elem : buttons_)
        {
                elem.update(input);
                if (elem.get_text() == "Start" && elem.is_triggered())
                {
                        return {StateSignal().set_push(new Game)};
                }
                else if ( elem.get_text() == "Quit" && elem.is_triggered())
                {
                        return {StateSignal().set_pop()};
                }
        }
        
        if (input.c == 'n')
        {
                return {StateSignal().set_push(new Game)};
        }

        if (input.c == 'q')
        {
                return {StateSignal().set_pop()};
        }

       
        
        return {};
}
